doc/buffer-watermark/align_watermark_flow_with_port_configuration.md

```diff
@@ -0,0 +1,285 @@
+                            gPortsOrch->addQueueWatermarkFlexCounters();
+                        }
+                        else if(key == PG_DROP_KEY)
+                        {
+                            gPortsOrch->generatePriorityGroupMap();
+                        }
+                        else if(key == PG_WATERMARK_KEY)
+                        {
+                            gPortsOrch->generatePriorityGroupMap();
+                            gPortsOrch->addPriorityGroupWatermarkFlexCounters();
+                        }
+```
+
+if queue or PG maps already created upon queue or pg-drop enablement and then watermark is enabled,
+the queue or PG maps won't be created again.
+this is done using the private boolean storing the created status. if it's already true, function returns.
+the same mechanism will be done in the new watermark functions.
+
+***PG map Example***
+
+```
+    if (m_isPriorityGroupMapGenerated)
+    {
+        return;
+    }
+```
+
+
+# 8. Testing
+
+## 8.1 Manual testing
+
+Test_flex_counter_logic(type=[watermark,queue,pg-drop]:
+1.	Disable the counter polling for all counters and save, reload configuration
+2.	Switch (type)
+    -   watermark: enable the watermark counter polling, check whether PG and queue map are generated in COUNTER_DB and both PG and QUEUE WATERMARK stats exists in FLEX_COUNTER_DB
+    -   queue: enable the queue counter polling, check whether queue map is generated in COUNTER_DB, no *WATERMARK* stats in FLEX_COUNTER_DB
+    -	pg-drop: enable the pg-drop counter polling, check whether PG map is generated in COUNTER_DB, no *WATERMARK* stats in FLEX_COUNTER_DB
+
+repeat this test 3 times, each time with different type [watermark,queue,pg-drop]
+Test_flex_counter_logic(type=[watermark,queue,pg-drop] reboot after counterpoll enable:
+1.	Disable the counter polling for all counters and save, reload configuration
+2.	Switch (type)
+    -   watermark: enable the watermark counter polling, check whether PG and queue map are generated in COUNTER_DB and both PG and QUEUE WATERMARK stats exists in FLEX_COUNTER_DB
+    -   queue: enable the queue counter polling, check whether queue map is generated in COUNTER_DB, no *WATERMARK* stats in FLEX_COUNTER_DB
+    -	pg-drop: enable the pg-drop counter polling, check whether PG map is generated in COUNTER_DB, no *WATERMARK* stats in FLEX_COUNTER_DB
+3. reboot switch
+   Switch (type)
+    -   watermark: watermark counter polling is enabled, check that PG and queue maps are generated in COUNTER_DB and both PG and QUEUE WATERMARK stats exists in FLEX_COUNTER_DB
+    -   queue: queue counter polling is enabled, check whether queue map is generated in COUNTER_DB, no *WATERMARK* stats in FLEX_COUNTER_DB
+    -	pg-drop: pg-drop counter polling is enabled, check whether PG map is generated in COUNTER_DB, no *WATERMARK* stats in FLEX_COUNTER_DB
+
+
+## 8.2 Regression testing
+
+sonic-mgmt/tests/qos/test_qos_sai.py on rpc image not the regular image (optional)

```

Comment:
Test cases in `test_qos_sai.py` are not relevant. They run on top of SAI API (via SAI thrift) and have nothing to do with Flexcounter.